File: c++/0139.cpp
// Link: https://leetcode.com/problems/word-break/description/
// Runtime: 15 ms
// Memory: 13 MB

class Solution 
{
public:
    bool wordBreak(const std::string& s, const std::vector<std::string>& wordDict) 
    {
        std::vector<bool> ending(s.size() + 1);
        ending[0] = true;

        for (int i = 1; i < static_cast<int>(s.size() + 1); i++)
        {
            int j = 0;
            bool isFound = false;
            while ((j < i) && (!isFound))
            {
                if ((ending[j]) && (std::find(wordDict.begin(), wordDict.end(), s.substr(j, i - j)) != wordDict.end()))
                {
                    ending[i] = true;
                    isFound = true;
                }

                j++;
            }
        }

        return ending[s.size()];
    }
};

File: python/0139.py
# Link: https://leetcode.com/problems/word-break/description/
# Runtime: 51 ms
# Memory: 16.45 MB

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        ending = [False] * (len(s) + 1)
        ending[0] = True

        for i in range(1, len(s) + 1):
            j, is_found = 0, False
            while (j < i) and (not is_found):
                if (ending[j]) and (s[j:i] in wordDict):
                    ending[i], is_found = True, True
                j += 1
        
        return ending[-1]
